src/mito2/src/memtable/time_series.rs

```diff
@@ -0,0 +1,840 @@
+
+            builder
+                .op_type
+                .extend_slice_of(&*v.op_type, 0, len)
+                .context(CompactValuesSnafu)?;
+
+            for (idx, f) in v.fields.iter().enumerate() {
+                builder.fields[idx]
+                    .extend_slice_of(&**f, 0, len)
+                    .context(CompactValuesSnafu)?;
+            }
+        }
+
+        let values = Values::from(builder);
+        self.frozen = vec![values.clone()];
+
+        Ok(values)
+    }
+}
+
+/// `ValueBuilder` holds all the vector builders for field columns.
+struct ValueBuilder {
+    region_metadata: RegionMetadataRef,
+    timestamp: Box<dyn MutableVector>,
+    sequence: UInt64VectorBuilder,
+    op_type: UInt8VectorBuilder,
+    fields: Vec<Box<dyn MutableVector>>,
+}
+
+impl ValueBuilder {
+    fn new(region_metadata: RegionMetadataRef, capacity: usize) -> Self {
+        let timestamp = region_metadata
+            .time_index_column()
+            .column_schema
+            .data_type
+            .create_mutable_vector(capacity);
+        let sequence = UInt64VectorBuilder::with_capacity(capacity);
+        let op_type = UInt8VectorBuilder::with_capacity(capacity);
+
+        let fields = region_metadata
+            .field_columns()
+            .map(|c| c.column_schema.data_type.create_mutable_vector(capacity))
+            .collect();
+
+        Self {
+            region_metadata,
+            timestamp,
+            sequence,
+            op_type,
+            fields,
+        }
+    }
+
+    /// Pushes a new row to `ValueBuilder`.
+    /// We don't need primary keys since they've already be encoded.
+    fn push(&mut self, ts: ValueRef, sequence: u64, op_type: u8, fields: Vec<ValueRef>) {
+        debug_assert_eq!(fields.len(), self.fields.len());
+        self.timestamp.push_value_ref(ts);
+        self.sequence.push_value_ref(ValueRef::UInt64(sequence));
+        self.op_type.push_value_ref(ValueRef::UInt8(op_type));
+        for (idx, field_value) in fields.into_iter().enumerate() {
+            self.fields[idx].push_value_ref(field_value);
+        }
+    }
+}
+
+/// [Values] holds an immutable vectors of field columns, including `sequence` and `op_typee`.
+#[derive(Clone)]
+struct Values {
+    timestamp: VectorRef,
+    sequence: Arc<UInt64Vector>,
+    op_type: Arc<UInt8Vector>,
+    fields: Vec<VectorRef>,
+}
+
+impl Values {
+    /// Sorts values in place by `timestamp, sequence, op_type`.
+    fn sort_in_place(&mut self) -> Result<()> {
+        let mut arrays = Vec::with_capacity(3 + self.fields.len());
+        arrays.push(self.timestamp.to_arrow_array());
+        arrays.push(self.sequence.to_arrow_array());
+        arrays.push(self.op_type.to_arrow_array());
+        arrays.extend(self.fields.iter().map(|f| f.to_arrow_array()));
+
+        let fields = arrays
+            .iter()
+            .map(|arr| datatypes::arrow::row::SortField::new(arr.data_type().clone()))
+            .collect();
```

Comment:
I think we only need to sort by `timestamp` and `sequence`. Here it uses all columns to create rows for sort.